lib/common/rules.c

```diff
@@ -888,3 +888,269 @@ pcmk__parse_source(const char *source)
+    }
+
+    /* Get name of node attribute to compare (expanding any %0-%9 to
+     * regular expression submatches)
+     */
+    attr = crm_element_value(expression, PCMK_XA_ATTRIBUTE);
+    if (pcmk__str_empty(attr)) {
+        pcmk__config_err("Treating " PCMK_XE_EXPRESSION " %s as not passing "
+                         "because " PCMK_XA_ATTRIBUTE " was not specified", id);
+        return pcmk_rc_unpack_error;
+    }
+    expanded_attr = pcmk__replace_submatches(attr, rule_input->rsc_id,
+                                             rule_input->rsc_id_submatches,
+                                             rule_input->rsc_id_nmatches);
+    if (expanded_attr != NULL) {
+        attr = expanded_attr;
+    }
+
+    // Get and validate operation
+    op = crm_element_value(expression, PCMK_XA_OPERATION);
+    comparison = pcmk__parse_comparison(op);
+    if (comparison == pcmk__comparison_unknown) {
+        // Not possible with schema validation enabled
+        if (op == NULL) {
+            pcmk__config_err("Treating " PCMK_XE_EXPRESSION " %s as not "
+                             "passing because it has no " PCMK_XA_OPERATION,
+                             id);
+        } else {
+            pcmk__config_err("Treating " PCMK_XE_EXPRESSION " %s as not "
+                             "passing because '%s' is not a valid "
+                             PCMK_XA_OPERATION, id, op);
+        }
+        rc = pcmk_rc_unpack_error;
+        goto done;
+    }
+
+    // How reference value is obtained (literal, resource meta-attribute, etc.)
+    source_s = crm_element_value(expression, PCMK_XA_VALUE_SOURCE);
+    source = pcmk__parse_source(source_s);
+    if (source == pcmk__source_unknown) {
+        // Not possible with schema validation enabled
+        // @COMPAT Fail expression once we can break backward compatibility
+        pcmk__config_warn("Expression %s has invalid " PCMK_XA_VALUE_SOURCE
+                          " value '%s', using default "
+                          "('" PCMK_VALUE_LITERAL "')", id, source_s);
+        source = pcmk__source_literal;
+    }
+
+    // Get and validate reference value
+    value = crm_element_value(expression, PCMK_XA_VALUE);
+    switch (comparison) {
+        case pcmk__comparison_defined:
+        case pcmk__comparison_undefined:
+            if (value != NULL) {
+                pcmk__config_warn("Ignoring " PCMK_XA_VALUE " in "
+                                  PCMK_XE_EXPRESSION " %s because it is unused "
+                                  "when " PCMK_XA_BOOLEAN_OP " is %s", id, op);
+                rc = pcmk_rc_unpack_error;
+                goto done;
+            }
+            break;
+
+        default:
+            if (value == NULL) {
+                pcmk__config_err("Treating " PCMK_XE_EXPRESSION " %s as not "
+                                 "passing because it has no " PCMK_XA_VALUE,
+                                 id);
+                rc = pcmk_rc_unpack_error;
```

Comment:
This is a behavioral change. I missed it on the first review. Current behavior with a missing value is not documented, but we make it to the comparison step.

```
$ sudo crm_attribute --query -n test_attr -N laptop -l reboot
scope=status  name=test_attr value=1

`eq` expression with missing value:
    <constraints>
      <rsc_location id="location-dummy" rsc="dummy">
        <rule id="location-dummy-rule" score="-INFINITY">
          <expression id="location-dummy-rule-expr" operation="eq" attribute="test_attr"/>
        </rule>
      </rsc_location>
    </constraints>

$ sudo crm_mon -1r | grep dummy
  * dummy	(ocf:pacemaker:Dummy):	 Started laptop

Change `operation` to `ne` and check again:
$ sudo crm_mon -1r | grep dummy
  * dummy	(ocf:pacemaker:Dummy):	 Stopped
```

Reply:
This is where the "before and after" unit tests might come in handy although it's a pain to backport the "after" ones

Reply:
good catch